src/llvm/types/di.rs

```diff
@@ -0,0 +1,455 @@
+/// debugging information ecosystem. It serves as a building block for more
+/// complex debug information entities such as scopes, types and variables.
+pub struct DINode<'ctx> {
+    pub(super) value_ref: LLVMValueRef,
+    _marker: PhantomData<&'ctx ()>,
+}
+
+impl<'ctx> DINode<'ctx> {
+    /// Constructs a new [`DINode`] from the given `value`.
+    ///
+    /// # Safety
+    ///
+    /// This method assumes that the provided `value` corresponds to a valid
+    /// instance of [LLVM `DINode`](https://llvm.org/doxygen/classllvm_1_1DINode.html).
+    /// It's the caller's responsibility to ensure this invariant, as this
+    /// method doesn't perform any validation checks.
+    pub(crate) unsafe fn from_value_ref(value_ref: LLVMValueRef) -> Self {
+        Self {
+            value_ref,
+            _marker: PhantomData,
+        }
+    }
+
+    /// Returns the low level `LLVMMetadataRef` corresponding to this node.
+    pub fn metadata(&self) -> LLVMMetadataRef {
+        unsafe { LLVMValueAsMetadata(self.value_ref) }
+    }
+
+    /// Returns a DWARF tag for the given debug info node.
+    pub fn tag(&self) -> DwTag {
+        DwTag(unsafe { LLVMGetDINodeTag(self.metadata()) })
+    }
+}
+
+impl<'ctx> TryFrom<DIDerivedType<'ctx>> for DINode<'ctx> {
+    type Error = ();
+
+    fn try_from(di_derived_type: DIDerivedType) -> Result<Self, Self::Error> {
+        // FIXME: perform a check
+        Ok(unsafe { Self::from_value_ref(di_derived_type.value_ref) })
+    }
+}
+
+impl<'ctx> TryFrom<DICompositeType<'ctx>> for DINode<'ctx> {
+    type Error = ();
+
+    fn try_from(di_composite_type: DICompositeType) -> Result<Self, Self::Error> {
+        // FIXME: perform a check
+        Ok(unsafe { Self::from_value_ref(di_composite_type.value_ref) })
+    }
+}
+
+/// Represents the debug information for a code scope.
+pub struct DIScope<'ctx> {
+    pub(super) value_ref: LLVMValueRef,
+    _marker: PhantomData<&'ctx ()>,
+}
+
+impl<'ctx> DIScope<'ctx> {
+    /// Constructs a new [`DIScope`] from the given `value`.
+    ///
+    /// # Safety
+    ///
+    /// This method assumes that the given `value` corresponds to a valid
+    /// instance of [LLVM `DIScope`](https://llvm.org/doxygen/classllvm_1_1DIScope.html).
+    /// It's the caller's responsibility to ensure this invariant, as this
+    /// method doesn't perform any validation checks.
+    pub(crate) unsafe fn from_value_ref(value_ref: LLVMValueRef) -> Self {
+        Self {
+            value_ref,
+            _marker: PhantomData,
+        }
+    }
+
+    /// Returns the low level `LLVMMetadataRef` corresponding to this node.
+    pub fn metadata(&self) -> LLVMMetadataRef {
+        unsafe { LLVMValueAsMetadata(self.value_ref) }
+    }
+
+    pub fn file(&self) -> DIFile {
+        unsafe {
+            let metadata = LLVMDIScopeGetFile(self.metadata());
+            DIFile::from_metadata_ref(metadata)
+        }
+    }
+}
+
+impl<'ctx> TryFrom<DICompositeType<'ctx>> for DIScope<'ctx> {
```

Comment:
DICompositeType is always a subclass of DIScope so this should use From not TryFrom right? It can't fail (this holds for all the other TryFrom that model the C++ class hierarchy)

Reply:
Right, fixed